mqtt/handler_test.go

```diff
@@ -0,0 +1,421 @@
+	handler := newHandler()
+	logBuffer.Reset()
+
+	chID, err := idProvider.ID()
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	var topics = []string{"channels/" + chID + "/messages"}
+
+	cases := []struct {
+		desc   string
+		client *session.Client
+		topic  []string
+		logMsg string
+	}{
+		{
+			desc:   "subscribe without active session",
+			client: nil,
+			topic:  topics,
+			logMsg: mqtt.LogErrFailedSubscribe + mqtt.ErrClientNotInitialized.Error(),
+		},
+		{
+			desc:   "subscribe with valid session and topics",
+			client: &sessionClient,
+			topic:  topics,
+			logMsg: fmt.Sprintf(mqtt.LogInfoSubscribed, clientID, topics[0]),
+		},
+	}
+
+	for _, tc := range cases {
+		handler.Subscribe(tc.client, &tc.topic)
+		assert.Contains(t, logBuffer.String(), tc.logMsg)
+	}
+}
+
+func TestUnsubscribe(t *testing.T) {
+	handler := newHandler()
+	logBuffer.Reset()
+
+	chID, err := idProvider.ID()
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	var topics = []string{"channels/" + chID + "/messages"}
+
+	cases := []struct {
+		desc   string
+		client *session.Client
+		topic  []string
+		logMsg string
+	}{
+		{
+			desc:   "unsubscribe without active session",
+			client: nil,
+			topic:  topics,
+			logMsg: mqtt.LogErrFailedUnsubscribe + mqtt.ErrClientNotInitialized.Error(),
+		},
+		{
+			desc:   "unsubscribe with valid session and topics",
+			client: &sessionClient,
+			topic:  topics,
+			logMsg: fmt.Sprintf(mqtt.LogInfoUnsubscribed, clientID, topics[0]),
+		},
+	}
+
+	for _, tc := range cases {
+		handler.Unsubscribe(tc.client, &tc.topic)
+		assert.Contains(t, logBuffer.String(), tc.logMsg)
+	}
+}
+
+func TestDisconnect(t *testing.T) {
+	handler := newHandler()
+	logBuffer.Reset()
+
+	chID, err := idProvider.ID()
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	var topics = []string{"channels/" + chID + "/messages"}
+
+	cases := []struct {
+		desc   string
+		client *session.Client
+		topic  []string
+		logMsg string
+	}{
+		{
+			desc:   "disconnect without active session",
+			client: nil,
+			topic:  topics,
+			logMsg: mqtt.LogErrFailedDisconnect + mqtt.ErrClientNotInitialized.Error(),
+		},
+		{
+			desc:   "disconnect with valid session",
+			client: &sessionClient,
+			topic:  topics,
+			logMsg: fmt.Sprintf(mqtt.LogInfoDisconnected, clientID, thingID),
+		},
+	}
+
+	for _, tc := range cases {
+		handler.Disconnect(tc.client)
+		assert.Contains(t, logBuffer.String(), tc.logMsg)
+	}
+}
+
+func newHandler() session.Handler {
+	logger, err := logger.New(&logBuffer, "debug")
+	if err != nil {
+		log.Fatalf("failed to create logger: %s", err)
+	}
+
+	authClient := mocks.NewClient(map[string]string{password: thingID}, map[string]interface{}{chanID: thingID})
+
+	var redisClient *redis.Client
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	container, err := pool.Run("redis", "5.0-alpine", nil)
```

Comment:
Spinning a new Docker container for each `newHandler` call is not a good idea. Why do we even use Redis and not mock for events?